src/main/java/org/folio/circulation/services/EventPublisher.java

```diff
@@ -56,6 +62,10 @@ public EventPublisher(RoutingContext routingContext) {
 
       return pubSubPublishingService.publishEvent(
         ITEM_CHECKED_OUT.name(), payloadJsonObject.encode())
+        .thenApply(r -> sendLogRecordEvent(new LogEventPayload()
+          .withLoggedObjectType(LogEventPayload.LoggedObjectType.CHECK_IN_CHECK_OUT)
+          .withAction(LogEventPayload.Action.CHECK_OUT)
+          .withBody(loanAndRelatedRecords.asJson().encode()), params))
```

Comment:
The publishing of the log record event seems to take a different approach rather than using the `pubSubPublishingService`, please could you help me understand why?

Reply:
Yes, we use different publisher implementation because it is some kind wrapper on PubSub. We considered case we can use another event bus implementation. So it will be simple to change logic in one place rather than make changes in different places in mod-circulation.

Reply:
> Yes, we use different publisher implementation because it is some kind wrapper on PubSub. 

What does this wrapper give that the existing PubSubPublishingService does not?

> We considered case we can use another event bus implementation. So it will be simple to change logic in one place rather than make changes in different places in mod-circulation.

The pub-sub publishing is already abstracted by the `pubSubPublishingService` and the `EventPublisher`. Are those seams insufficient for potential replacement of the event publishing?

Reply:
@VasilyGan, could you please help us?

Reply:
@marcjohnson-kint, @TarasSpashchenko is creating the new client library for Kafka. It should be more efficient in terms of performance than PubSub and it won't be part of the PubSub (@TarasSpashchenko, please, correct if something wrong here). The initial version of the Audit log solution should leverage PubSub. Once the new library is released, it makes sense to switch to it. To make the migration easier, it is better to create an abstract layer (library), which should allow to send Audit log messages and encapsulate the concrete implementation of the sending mechanism.

Reply:
> @TarasSpashchenko is creating the new client library for Kafka. 

He is. As I understand it, it has not yet been decided if general FOLIO modules will have direct access to Kafka.

> The initial version of the Audit log solution should leverage PubSub. Once the new library is released, it makes sense to switch to it. To make the migration easier, it is better to create an abstract layer (library), which should allow to send Audit log messages and encapsulate the concrete implementation of the sending mechanism.

Sure. What I don't understand is why the current abstractions are insufficient. In theory, we would only need to replace the pub-sub publishing service with a Kafka publishing service. What would an additional abstraction gain us?

Reply:
@marcjohnson-kint, to be more specific, how would you change this solution? Can you send a code snippet to illustrate it, please? And why it would make it better than the current approach proposed in this PR?

Reply:
@VasilyGan, there  are already some abstractions in circulation that allows communication with pub/sub.

See `PubSubPublishingService` and `EventPublisher`. I'd also prefer to reuse this classes instead of introducing dependency on the mod-audit-client. Here is `org.folio.circulation.services.EventPublisher#publishItemCheckedOutEvent` method that used to publish item checked out event, which consumed by patron blocks module.

Reply:
Since the new library will be also used by other modules, it is better to add it to the circulation, so that a way Audit log messages are sent will be unified. 
PubSubPublishingService and EventPublisher are part of the circulation module and can't be re-used by other modules, unfortunately.

Reply:
@marcjohnson-kint, if you still have questions regarding this library, let's setup a meeting and discuss them. I think it will be faster than continue this thread. 
Or, if you agree with the last comment, let's keep the current approach.

Reply:
> Since the new library will be also used by other modules, it is better to add it to the circulation, so that a way Audit log messages are sent will be unified.
> PubSubPublishingService and EventPublisher are part of the circulation module and can't be re-used by other modules, unfortunately.

An alternative could have been to extract the implementations from mod-circulation out, if it was desirable to reuse them.

I get the sense that you are very keen for the audit specific library to be used because it is already written.

Reply:
I can confirm that it is rather simple to switch on mod-circulation publisher at the present moment. If @VasilyGan hasn't objections on this I can easily provide this change.

Reply:
> if you still have questions regarding this library, let's setup a meeting and discuss them. I think it will be faster than continue this thread.

I still don't really understand the need for it, over either the modules current way of publishing to pub-sub or using the existing pub-sub client library. That is likely because I'm missing context.

I was specifically asked to provide a review for this change by Victoria because the FireBird team wanted an external review. Folks are free to disregard my feedback, it is entirely optional.

I've already expressed reservations about me reviewing this work because of my general concerns with the approach and the alignment of this work with the proposal.

@ VasilyGan @khandramai Maybe it would be more appropriate to ask someone else to provide that external review?

Reply:
@marcjohnson-kint, your opinion is always welcome. If you want to discuss this feature architecture one more time or the reason why the new library is used here, I'm happy to discuss it with you via Slack or during a meeting. 
Thank you.

Reply:
> @marcjohnson-kint, your opinion is always welcome. 

Thank you.

> If you want to discuss this feature architecture one more time or the reason why the new library is used here, I'm happy to discuss it with you via Slack or during a meeting.

Thank you for the invitation.

I am away next week. Given how the implementation has evolved in relation to the proposed design, I'll reflect on what feedback might be useful when I return.